# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/builtin/FinalizationRegistryObject.cpp
# Commit: 5a6684922b98
# Full Hash: 5a6684922b980d32b05bd7210cfadc34e734a845
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2020-03-21 21:28:09
# Regressor Bug: 1623973
# File Overlap Count: 2
# Description:
#   Bug 1623973 - Make FinalizationRegistryObject active record set and registrations map values weak r=sfink
#   
#   This removes tracing for the values of the registration map and the active record set and sweeps them instead. This requires maintaining a map of FinalizationRegistryObjects on the zone.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D67677
# ==============================================================================

diff -r c4d2ca8f78b7 -r 5a6684922b98 js/src/builtin/FinalizationRegistryObject.cpp
--- a/js/src/builtin/FinalizationRegistryObject.cpp	Thu Mar 19 16:42:42 2020 +0000
+++ b/js/src/builtin/FinalizationRegistryObject.cpp	Fri Mar 20 21:57:22 2020 +0000
@@ -151,7 +151,7 @@
     nullptr,                                   // call
     nullptr,                                   // hasInstance
     nullptr,                                   // construct
-    FinalizationRecordVectorObject::trace,     // trace
+    nullptr,                                   // trace
 };
 
 /* static */
@@ -175,14 +175,6 @@
 }
 
 /* static */
-void FinalizationRecordVectorObject::trace(JSTracer* trc, JSObject* obj) {
-  auto rv = &obj->as<FinalizationRecordVectorObject>();
-  if (auto* records = rv->records()) {
-    records->trace(trc);
-  }
-}
-
-/* static */
 void FinalizationRecordVectorObject::finalize(JSFreeOp* fop, JSObject* obj) {
   auto rv = &obj->as<FinalizationRecordVectorObject>();
   fop->delete_(obj, rv->records(), MemoryUse::FinalizationRecordVector);
@@ -224,6 +216,11 @@
   records()->eraseIfEqual(record);
 }
 
+inline void FinalizationRecordVectorObject::sweep() {
+  MOZ_ASSERT(records());
+  return records()->sweep();
+}
+
 ///////////////////////////////////////////////////////////////////////////
 // FinalizationRegistryObject
 
@@ -329,6 +326,10 @@
   registry->initReservedSlot(IsQueuedForCleanupSlot, BooleanValue(false));
   registry->initReservedSlot(IsCleanupJobActiveSlot, BooleanValue(false));
 
+  if (!cx->runtime()->gc.addFinalizationRegistry(cx, registry)) {
+    return false;
+  }
+
   args.rval().setObject(*registry);
   return true;
 }
@@ -336,14 +337,37 @@
 /* static */
 void FinalizationRegistryObject::trace(JSTracer* trc, JSObject* obj) {
   auto registry = &obj->as<FinalizationRegistryObject>();
+
+  // Trace the registrations weak map. At most this traces the
+  // FinalizationRecordVectorObject values of the map; the contents of those
+  // objects are weakly held and are not traced.
   if (ObjectWeakMap* registrations = registry->registrations()) {
     registrations->trace(trc);
   }
-  if (FinalizationRecordSet* records = registry->activeRecords()) {
+
+  // The active record set is weakly held and is not traced.
+
+  if (FinalizationRecordVector* records = registry->recordsToBeCleanedUp()) {
     records->trace(trc);
   }
-  if (FinalizationRecordVector* records = registry->recordsToBeCleanedUp()) {
-    records->trace(trc);
+}
+
+void FinalizationRegistryObject::sweep() {
+  // Sweep the set of active records. These may die if CCWs to record objects
+  // get nuked.
+  MOZ_ASSERT(activeRecords());
+  activeRecords()->sweep();
+
+  // Sweep the contents of the registrations weak map's values.
+  MOZ_ASSERT(registrations());
+  for (ObjectValueWeakMap::Enum e(registrations()->valueMap()); !e.empty();
+       e.popFront()) {
+    auto registrations =
+        &e.front().value().toObject().as<FinalizationRecordVectorObject>();
+    registrations->sweep();
+    if (registrations->isEmpty()) {
+      e.removeFront();
+    }
   }
 }
 
@@ -354,7 +378,7 @@
   // Clear the weak pointer to the registry in all remaining records.
 
   // FinalizationRegistries are foreground finalized whereas record objects are
-  // background finalized, so record objects and guaranteed to still be
+  // background finalized, so record objects are guaranteed to still be
   // accessible at this point.
   MOZ_ASSERT(registry->getClass()->flags & JSCLASS_FOREGROUND_FINALIZE);
 